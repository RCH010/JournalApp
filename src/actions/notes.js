import Swal from 'sweetalert2';
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";


// con lo de thunk, que es lo que ayuda con el middleware para las acciones asincronas
//  tiene además del dispatch, el getState, y se puede acceder a esos, y ya del
//  get state, podemos acceder a los valores del state, o con el dispatch para
//  ejecutar una acción y se modifique algo del state
export const startNewNote = () => {
    return async (dispatch, getState) => {
        const {uid} = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);
        dispatch(activeNote(docRef.id, newNote));
        dispatch(addNewNote(docRef.id, newNote));
    }
}
// Action to set a note in the state
export const activeNote = (id, note) => (
    {
        type: types.notesActive,
        payload: {
            id,
            ...note
        }
    }
)

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}
 
export const setNotes = (notes) => ({
    type: types.noteLoad,
    payload: notes
});

// action to update notes in firestore
export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const {uid} = getState().auth;

        
        const noteToFirestore = {...note};
        delete noteToFirestore.id;

        if(!noteToFirestore.url) { delete noteToFirestore.url; }
        if(!noteToFirestore.body) { delete noteToFirestore.body; }

        await db.doc(`/${uid}/journal/notes/${note.id}`).update(noteToFirestore)
            .then(() => {
                dispatch(refreshNote(note.id, note));
                Swal.fire('Saved', note.title, 'success');
            })
            .catch(() => {
                Swal.fire('Something went wrong :(', note.title, 'error');
            })
    }
}

// action to update the note that's just been updated
//  so in the list of notes it is updated
export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id, 
        note: {
            id,
            ...note
        }
    }
})

// action that calls ./helpers/fileUpload to upload the image and returns
//  image url, then we call action startSaveNote so the db is updated
export const startUploadingPicture = (file) => {
    return async (dispatch, getState) => {
        const {active: activeNote} = getState().notes;
        // UI feedback
        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait.',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            }
        }); 
        const fileUrl = await fileUpload(file);
        
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote));

        Swal.close();
    }
}

//
export const startDeleting = (noteId) => {
    return async (dispatch, getState) => {
        const {uid} = getState().auth;
        // deleted from firestore
        await db.doc(`${uid}/journal/notes/${noteId}`).delete();

        //now deleted from store
        dispatch(deleteNote(noteId));
    }
}

export const deleteNote = (noteId) => ({
    type: types.notesDelete,
    payload: noteId,
})

export const notesLogout = () => ({
    type: types.notesLogoutCleaning,
})
