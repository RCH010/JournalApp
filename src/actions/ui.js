import { types } from "../types/types"

export const setError = (errorMsg) => {
    return {
        type: types.uiSetError,
        payload: errorMsg
    };
};

export const removeError = () => {
    return {
        type: types.uiRemoveError
    };
};

export const startLoading = () => {
    return {
        type: types.uiStartLoading
    }
};

export const finishLoading = () => {
    return {
        type: types.uiFinishLoading
    };
}