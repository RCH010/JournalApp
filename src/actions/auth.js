import { googleAuthProvider, firebase } from "../firebase/firebaseConfig";
import { types } from "../types/types"
import { finishLoading, startLoading } from "./ui";

//  CUANDO ES UNA TAREA ASINCRONA, SE NECESITA RETORNAR UN CALLBACK
export const startLoginWithEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());
        firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName, user.email)
                );
            })
            .catch(e => {
                console.log(e);
            })
            .finally(() => {
                dispatch(finishLoading())
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then( async({user}) => {
                await user.updateProfile({displayName: name});
                dispatch(
                    login(user.uid, user.displayName, user.email)
                )
            })
            .catch( e => {
                console.log(e);
            })
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName, user.email)
                )
            });
    }
}

export const login = (uid, displayName, email) => ( {
    type: types.login,
    payload: {
        uid,
        displayName,
        email
    }
})