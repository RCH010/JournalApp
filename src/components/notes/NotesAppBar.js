import React from 'react'
import moment from 'moment';
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploadingPicture } from '../../actions/notes'

export const NotesAppBar = () => {
    
    const dispatch = useDispatch();
    const {active} = useSelector(state => state.notes)

    const handleSave = () => {
        dispatch(startSaveNote(active));
    }

    const handlePictureUpload = () => {
        document.querySelector('#fileSelector').click(); 
    }
    const handleFileChange = (e) => {
        const file = e.target.files[0];
        
        if(file) {
            dispatch(startUploadingPicture(file));
        }
    }

    return (
        <div className='notes__app-bar'>
            <span>{moment().format('LL')}</span>
            <input 
                id='fileSelector'
                type='file'
                style={{display: 'none'}}
                onChange={handleFileChange}
            />
            <div>
                <button  
                    className='btn'
                    onClick={handlePictureUpload}
                >
                    Add Picture
                </button>
                <button 
                    className='btn'
                    onClick={handleSave}
                >
                    Save
                </button>
            </div>
        </div>
    )
}
