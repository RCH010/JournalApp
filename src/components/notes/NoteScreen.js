import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote } from '../../actions/notes'
import { useForm } from '../../hooks/useForm'
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {
    
    const dispatch = useDispatch();
    const {active: note} = useSelector(state => state.notes)
    const [formValues, handleInputChange, reset] = useForm(note);
    const {body, title} = formValues;

    const activeId = useRef(note.id)

    useEffect(() => {
        if(note.id !== activeId.current){
            reset(note);
            activeId.current = note.id;
        }
    }, [note, reset ])

    useEffect(() => {
        dispatch(activeNote(formValues.id, {...formValues}));        
    }, [formValues, dispatch]);

    return (
        <div className='notes__main-content'>
            <NotesAppBar />
            <div className='notes__content'>

                <input
                    type='text'
                    name='title'
                    placeholder='Title'
                    className='notes__title-input'
                    autoComplete='off'
                    value={title}
                    onChange={handleInputChange}
                />
                <textarea
                    placeholder="What's new?"
                    className='notes__textarea mt-1'
                    name='body'
                    value={body}
                    onChange={handleInputChange}
                >
                </textarea>
                {
                    (note.url) &&
                    (<div 
                        className='notes__image'
                    >
                        <img 
                            src={note.url} 
                            alt='imagen'
                        />
                    </div>)
                }
            </div>
        </div>
    )
}
