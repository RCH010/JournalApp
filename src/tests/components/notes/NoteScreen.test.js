import { mount } from "enzyme"
import thunk from "redux-thunk";
import configureStore from 'redux-mock-store';
import { NoteScreen } from "../../../components/notes/NoteScreen"
import { Provider } from "react-redux";
import { activeNote } from "../../../actions/notes";
import { act } from "react-dom/test-utils";

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {
        uid: '1234ABC',
        name: 'Testing',
        email:'test@tesing.com'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: {
            id: '12344',
            body: 'Wenas',
            title: 'Ya soy titulo',
        }
    }
};

let store = mockStore(initState)
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
            <NoteScreen />
    </Provider>
)
describe('Tests in <NoteScreen/>', () => {


    test('should match NoteScreen with Snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should execute activeNote action', () => {
        act( ()=> {
            wrapper.find({name:'title'}).prop('onChange')({target: {name: 'title', value: 'I am a Title'}});
        })
        expect(activeNote).toHaveBeenLastCalledWith("12344", {"body": "Wenas", "id": "12344", "title": "I am a Title"});
    })
    
})
