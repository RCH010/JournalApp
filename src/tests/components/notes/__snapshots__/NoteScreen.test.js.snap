// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <NoteScreen/> should match NoteScreen with Snapshot 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__app-bar"
  >
    <span>
      February 17, 2021
    </span>
    <input
      id="fileSelector"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Add Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Title"
      type="text"
      value="Ya soy titulo"
    />
    <textarea
      className="notes__textarea mt-1"
      name="body"
      onChange={[Function]}
      placeholder="What's new?"
      value="Wenas"
    />
  </div>
  <div
    className="notes__buttons-area mb-1"
  >
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Delete
    </button>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Save
    </button>
  </div>
</div>
`;
